parsec/core/fs/workspacefs/workspacefs.py

```diff
@@ -440,6 +441,193 @@ def get_encryption_revision(self) -> int:
+
+                async def _populate_path_w_index(data, index, entry_id, timestamp):
+                    try:
+                        data[index] = await _get_past_path(entry_id, timestamp=timestamp)
+                    except (FSRemoteManifestNotFoundBadVersion, FSEntryNotFound):
+                        pass
+
+                # TODO : Use future manifest source field to follow files and directories
+                async with trio.open_nursery() as child_nursery:
+                    child_nursery.start_soon(
+                        _populate_path_w_index, data, 1, entry_id, early.add(microseconds=-1)
+                    )
+                    child_nursery.start_soon(_populate_path_w_index, data, 2, entry_id, late)
+                    child_nursery.start_soon(_populate_path_w_index, data, 3, entry_id, early)
+                tree[(manifest.id, manifest.version, early, late)] = (
+                    data[0],
+                    data[1] if data[1] != data[3] else None,
+                    data[2] if data[2] != data[3] else None,
+                )
+            else:
+                if not is_file_manifest(manifest):
+                    for child_name, child_id in manifest.children.items():
+                        if child_name == target.parts[path_level + 1]:
+                            return await _populate_tree_list_versions(
+                                nursery,
+                                target,
+                                path_level + 1,
+                                tree,
+                                child_id,
+                                early if early > manifest.updated else manifest.updated,
+                                late,
+                            )
+                else:
+                    pass  # TODO : Broken path. What to do?
+
+        async def _populate_tree_list_versions(
+            nursery,
+            target: FsPath,
+            path_level: int,
+            tree: dict,
+            entry_id: EntryID,
+            early: Pendulum,
+            late: Pendulum,
+        ):
+            # TODO : Check if directory, melt the same entries through different parent
+            versions = await self.remote_loader.list_versions(entry_id)
+            for version, (timestamp, creator) in versions.items():
+                next_version = min((v for v in versions if v > version), default=None)
+                nursery.start_soon(
+                    _populate_tree_load,
+                    nursery,
+                    target,
+                    path_level,
+                    tree,
+                    entry_id,
+                    max(early, timestamp),
+                    late if next_version not in versions else min(late, versions[next_version][0]),
+                    version,
+                    next_version,
+                )
+
+        return_tree = {}
+        root_manifest = await self.transactions._get_manifest_from_path(FsPath("/"))
```

Comment:
`workspace_manifest = await self.local_storage.get_manifest(self.workspace_id)`

Reply:
Couldn't we imagine some special case where the root manifest has been garbage collected? I wasn't sure... I will change this anyway, and put the whole algorithm in a new module as it's indeed a big enough piece of code

Reply:
>  Couldn't we imagine some special case where the root manifest has been garbage collected? I wasn't sure...

Oh right. Then:

    workspace_manifest = await self.transactions._get_manifest(self.workspace_id)